mysql-persistence/src/main/java/com/netflix/conductor/dao/mysql/MySQLQueueDAO.java

```diff
@@ -0,0 +1,241 @@
+			Number queueSize = (Number)row.get("size");
+			detail.put(queueName, queueSize.longValue());
+		}));
+		return detail;
+	}
+
+	@Override
+	public Map<String, Map<String, Map<String, Long>>> queuesDetailVerbose() {
+		Map<String, Map<String, Map<String, Long>>> result = Maps.newHashMap();
+
+		withTransaction(tx -> tx.createQuery(GET_QUEUES_DETAIL).executeAndFetchTable().asList().forEach(row -> {
+			String queueName = (String)row.get("queue_name");
+			Number queueSize = (Number)row.get("size");
+			Number queueUnacked = (Number)row.get("uacked");
+			result.put(queueName, ImmutableMap.of(
+					"a", ImmutableMap.of(   //sharding not implemented, returning only one shard with all the info
+						"size", queueSize.longValue(),
+						"uacked", queueUnacked.longValue()
+					)
+				)
+			);
+		}));
+
+		return result;
+	}
+
+	@Override
+	public void processUnacks(String queueName) {
+
+	}
+
+	private boolean existsMessage(Connection connection, String queueName, String id) {
+		return connection.createQuery(EXISTS_MESSAGE)
+				.addParameter("queueName", queueName)
+				.addParameter("messageId", id)
+				.executeScalar(Boolean.class);
+	}
+
+	private void pushMessage(Connection connection, String queueName, String messageId, String payload, long offsetTimeInSecond) {
+		createQueueIfNotExists(connection, queueName);
+		Date now = DateUtils.truncate(new Date(), Calendar.SECOND);
+		Date deliverTime = new Date(now.getTime() + (offsetTimeInSecond*1000));
+		boolean exists = connection.createQuery(EXISTS_MESSAGE).addParameter("queueName", queueName).addParameter("messageId", messageId).executeScalar(Boolean.class);
+		if (!exists) {
+			connection.createQuery(PUSH_MESSAGE)
+					.addParameter("createdOn", now)
+					.addParameter("deliverOn", deliverTime)
+					.addParameter("queueName", queueName)
+					.addParameter("messageId", messageId)
+					.addParameter("offsetSeconds", offsetTimeInSecond)
+					.addParameter("payload", payload).executeUpdate();
+		} else {
+			connection.createQuery(UPDATE_MESSAGE)
+					.addParameter("queueName", queueName)
+					.addParameter("messageId", messageId)
+					.addParameter("payload", payload).executeUpdate();
+		}
+	}
+
+	private void removeMessage(Connection connection, String queueName, String messageId) {
+		connection.createQuery(REMOVE_MESSAGE).addParameter("queueName", queueName).addParameter("messageId", messageId).executeUpdate();
+	}
+
+	private List<String> peekMessages(String queueName, int count) {
+		if (count < 1) return Collections.emptyList();
+		return getWithTransaction(tx -> tx.createQuery(PEEK_MESSAGES)
+				.addParameter("queueName", queueName)
+				.addParameter("count", count)
+				.executeScalarList(String.class));
+	}
+
+	private List<String> popMessages(Connection connection, String queueName, List<String> messageIds) {
+		if (messageIds.isEmpty()) return messageIds;
+		List<String> stringIds = messageIds.stream().map(id -> format("'%s'", id)).collect(Collectors.toList());
```

Comment:
This seems a little dangerous. Assuming it has been verified that all of the messageId strings are clean, a comment should be added here explaining why this is safe.

But, considering the proliferation of IN clauses and sql2o's [lack of support](https://github.com/aaberg/sql2o/issues/146), I'm not sure how good a fit that library is.

Reply:
Good point @tommack, I am a bit uneasy with the lack of `IN` support as well. The thing is I didn't find that many libraries over JDBC that provide a convenient programming model and that are performant.

I don't think we want to bring in Hibernate/JPA for this right? Seems a bit heavy.

Do you know any well performing library that could replace Sql2o. I could consider switching the lib ... I want to have the best possible implementation.

Other alternative would be to add input sanitization logic and comments as you suggest.

Reply:
@tommack I found a better to do this still with Sql2o. The Query api has a withParams method that supports an array and expects the query to have a list of p1, p2, pN placeholders

So instead I will just generate a dynamic list of placeholders of the query and then use that API. That will ensure that the PreparedStatement can properly escape any dangerous SQL